plugins/woocommerce/src/StoreApi/Routes/V1/CartUpdateCustomer.php

```
@@ -213,6 +213,24 @@ protected function get_route_post_response( \WP_REST_Request $request ) {
 
 		$customer->save();
 
+		// Fix for race condition: Update the draft order directly with fresh customer data
+		// This prevents stale session data from overwriting recent updates when concurrent
+		// requests occur (e.g., batch POST followed immediately by GET checkout).
+		$draft_order = $this->get_draft_order();
+
+		// If no draft order exists yet, create one so the data is properly stored.
+		if ( ! $draft_order && ! WC()->cart->is_empty() ) {
+			$draft_order = $this->order_controller->create_order_from_cart();
```

In other areas we add some logging after we create an order, see https://github.com/woocommerce/woocommerce/blob/56b670b5c2975e424d3dea12eca84c292d555ad0/plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php#L647

---

`create_order_from_cart` can throw an exception, so it would be good to add some error handling here.

---

> In other areas we add some logging after we create an order, see

Good idea, added.

> `create_order_from_cart` can throw an exception, so it would be good to add some error handling here.

The exception thrown in `create_order_from_cart` would be handled by the REST API and an error response would be returned, catching locally would suppress the error

---

Has the log been useful? Just wondering how much overhead it adds to requests...